cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/TaggedInputSplit.java

```diff
@@ -0,0 +1,197 @@
+  private Map<String, String> inputConfigs;
+
+  public TaggedInputSplit() {
+  }
+
+  /**
+   * Creates a new TaggedInputSplit.
+   *
+   * @param name name of the InputSplit
+   * @param inputSplit The InputSplit to be tagged
+   * @param conf The configuration to use
+   * @param inputConfigs configurations to use for the InputFormat
+   * @param inputFormatClass The InputFormat class to use for this job
+   * @param mapperClass The Mapper class to use for this job
+   */
+  @SuppressWarnings("unchecked")
+  public TaggedInputSplit(String name, InputSplit inputSplit, Configuration conf,
+                          Map<String, String> inputConfigs,
+                          Class<? extends InputFormat> inputFormatClass,
+                          Class<? extends Mapper> mapperClass) {
+    this.name = name;
+    this.inputSplitClass = inputSplit.getClass();
+    this.inputSplit = inputSplit;
+    this.conf = conf;
+    this.inputConfigs = inputConfigs;
+    this.inputFormatClass = (Class<? extends InputFormat<?, ?>>) inputFormatClass;
+    this.mapperClass = mapperClass;
+  }
+
+  /**
+   * Retrieves the name of this TaggedInputSplit.
+   *
+   * @return name of the TaggedInputSplit
+   */
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * Retrieves the original InputSplit.
+   *
+   * @return The InputSplit that was tagged
+   */
+  public InputSplit getInputSplit() {
+    return inputSplit;
+  }
+
+  /**
+   * Retrieves the InputFormat class to use for this split.
+   *
+   * @return The InputFormat class to use
+   */
+  @SuppressWarnings("unchecked")
+  public Class<? extends InputFormat<?, ?>> getInputFormatClass() {
+    return inputFormatClass;
+  }
+
+  /**
+   * Retrieves the Mapper class to use for this split.
+   *
+   * @return The Mapper class to use
+   */
+  @SuppressWarnings("unchecked")
+  public Class<? extends Mapper> getMapperClass() {
+    return mapperClass;
+  }
+
+  @Override
+  public long getLength() throws IOException, InterruptedException {
+    return inputSplit.getLength();
+  }
+
+  @Override
+  public String[] getLocations() throws IOException, InterruptedException {
+    return inputSplit.getLocations();
+  }
+
+  Map<String, String> getInputConfigs() {
+    return inputConfigs;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void readFields(DataInput in) throws IOException {
+    name = Text.readString(in);
+    inputSplitClass = (Class<? extends InputSplit>) readClass(in);
+    inputFormatClass = (Class<? extends InputFormat<?, ?>>) readClass(in);
+    mapperClass = (Class<? extends Mapper<?, ?, ?, ?>>) readClass(in);
+    inputConfigs = GSON.fromJson(Text.readString(in), STRING_STRING_MAP_TYPE);
+    inputSplit = ReflectionUtils.newInstance(inputSplitClass, conf);
+    SerializationFactory factory = new SerializationFactory(conf);
+    Deserializer deserializer = factory.getDeserializer(inputSplitClass);
+    deserializer.open((DataInputStream) in);
```

Comment:
I tried closing the serializer/deserialize, but then I get this when running the MR job:
![image](https://cloud.githubusercontent.com/assets/2440977/13731136/4f0d1ff8-e91f-11e5-9fb6-a9f937dda68e.png)

Calling close on the serializer/deserializer closes underlying InputStream/OutputStream, and I don't think we're supposed to be responsible for closing it, since we're passed it in the method.